code/messaging/src/phone/java/com/adobe/marketing/mobile/MessagingInternal.java

```diff
@@ -471,6 +502,221 @@ private static void addXDMData(final EventData eventData, final Map<String, Obje
+                return;
+            }
+            final JSONObject items = payload.getJSONArray(MessagingConstant.EventDataKeys.Offers.ITEMS).getJSONObject(0);
+            rulesJsonString = items.getJSONObject(MessagingConstant.EventDataKeys.Offers.DATA).getString(MessagingConstant.EventDataKeys.Offers.CONTENT);
+        } catch (JSONException e) {
+            Log.debug(LOG_TAG, "handleOfferNotification - JSON exception when attempting to retrieve rules json from the edge response event: %s", e.getLocalizedMessage());
+        }
+
+        final JsonUtilityService.JSONObject rulesJsonObject = getJsonUtilityService().createJSONObject(rulesJsonString);
+        final  List<Rule> rulesList = parseRulesFromJsonObject(rulesJsonObject);
+        messagingModule.replaceRules(rulesList);
+    }
+
+    /**
+     * Parses all rules from the provided {@code jsonObject} into a list of {@code Rules}s.
+     * <p>
+     * If input {@code jsonObject} is null or, if there is an error reading {@link JsonUtilityService.JSONArray} from it,
+     * empty {@code List<Rule>} is returned.
+     *
+     * @param jsonObject {@code JSONObject} containing the list of rules and consequences
+     * @return a {@code List} of {@code Rule} objects that were parsed from the input {@code jsonObject}
+     */
+    private List<Rule> parseRulesFromJsonObject(final JsonUtilityService.JSONObject jsonObject) {
+        final List<Rule> parsedRules = new ArrayList<>();
+
+        if (jsonObject == null) {
+            Log.debug(LOG_TAG, "parseRulesFromJsonObject -  Unable to parse rules, input jsonObject is null.");
+            return parsedRules;
+        }
+
+        JsonUtilityService.JSONArray rulesJsonArray;
+
+        try {
+            rulesJsonArray = jsonObject.getJSONArray(MessagingConstant.EventDataKeys.RulesEngine.JSON_KEY);
+        } catch (final JsonException e) {
+            Log.debug(LOG_TAG, "parseRulesFromJsonObject -  Unable to parse rules (%s)", e);
+            return parsedRules;
+        }
+
+        // loop through each rule definition
+        for (int i = 0; i < rulesJsonArray.length(); i++) {
+            try {
+                // get individual rule json object
+                final JsonUtilityService.JSONObject ruleObject = rulesJsonArray.getJSONObject(i);
+                // get rule condition
+                final JsonUtilityService.JSONObject ruleConditionJsonObject = ruleObject.getJSONObject(
+                        MessagingConstant.EventDataKeys.RulesEngine.JSON_CONDITION_KEY);
+                final RuleCondition condition = RuleCondition.ruleConditionFromJson(ruleConditionJsonObject);
+                // get consequences
+                final List<Event> consequences = generateConsequenceEvents(ruleObject.getJSONArray(
+                        MessagingConstant.EventDataKeys.RulesEngine.JSON_CONSEQUENCES_KEY));
+                if(consequences != null) {
+                    parsedRules.add(new Rule(condition, consequences));
```

Comment:
Do we need to put null check on condition?

Reply:
yes otherwise its possible for a rule to be added which has no consequence.